voxeltycoon-community/voxel-tycoon-open-library
# CX - Automatic Project Building and Releasing

# CX - Automatic Project Building and Releasing



**Goal**: CX stands for Continuous X and it is the umbrella term to combine different software processes such as [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration) and [Continues Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) into a single automated program. This program will be responsible for:

- Style Checking our code.
- Compiling our code.
- Running tests on our code.
- Creating releases and deploying our code.

Having a CX program is standard in software development and can be accomplished for us through the use of [GitHub Actions](https://github.com/features/actions). However, this program does require access to the Voxel Tycoon dlls so that it can properly compile the program. Extra consideration should be taken to provide these dlls in a secure manner for the CX without exposing them to the public (accidentally encouraging pirating and cracked games).

## How to Test





The CX will only run when its corresponding file is on the master/main branch. Therefore, no testing can be done on the PR itself.

When this file is present then each time a change is pushed (let it be to master/main or on a pull request) then this will automatically build and test the project. When the everything works as expected and the push was made on the master/main branch then this will also automatically deploy our project/mod to Steam!

## TODO list



- [x] Configure Github Actions Workflow File.
- [x] Ask andrewpey about the Voxel Tycoon `.dll`s for building.
- [x] Give the CX bot access to the `.dll`s via the the workflow file.
    - [x] Through github actions let the bot download the dll project.
    - [x] Set the environment variables so that our project can find the dlls.
    - [x] Specify the secrets to access the private dll project.
        - [x] `data_repo` specifies the URL to access the dll repository.
        - [x] `data_branch` specifies the branch in the repository to pull from.
        - [x] `data_token` specifies the access token to enter this repository.
        - [x] `data_path` specifies the name of the folder this repo should be installed in (following the VT convention). 
        - [x] `steam_gameid` specifies the id of Voxel Tycoon on steam.
- [x] Let the CX bot auto-publish our game.
    - [x] Create a Steam account for the bot to post our mod.
    - [x] Buy the game for the bot so they can publish the mod.
    - [x] Create the initial mod, manually, for the bot to post the game. 
    - [x] Specify the secrets to automatically publish the mod. 
        - [x] `steam_workshopid` specified the id of our mod as on the steam workshop.
        - [x] `steam_username` specifies the user name of the account holding the VTOL mod.
        - [x] `steam_password` specifies the password of the account holding the VTOL mod.
        - [x] ~~`steam_tfaseed` specifies the two factor authentication seed of the account holding the VTOL mod.~~ Failed to retrieve the 2FA seed.
- [x] Upload and test on the repo.

## Significant Changes




- Created the [continuous deployment][continuous-deployment.yml] file, handling the Github Actions workflow to build, test and publish this project.





## Relevant Issues




closes #8




[continuous-deployment.yml]: https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/continuous-deployment/.github/workflows/continuous-deployment.yml
